pkp/pkp-lib#7135 Multiple author affiliations

GaziYucel opened the pull request:
Related issues:
- https://github.com/pkp/pkp-lib/issues/7135

Pull requests:
- https://github.com/pkp/pkp-lib/pull/10460
- https://github.com/pkp/ui-library/pull/417
- https://github.com/pkp/ojs/pull/4469
- https://github.com/pkp/omp/pull/1819
- https://github.com/pkp/googleScholar/pull/23
- https://github.com/pkp/jatsTemplate/pull/68
- https://github.com/asmecher/grobidMetadataExtractor/pull/1
- https://github.com/pkp/quickSubmit/pull/102
- https://github.com/pkp/crossref-ojs/pull/59

GaziYucel commented:
Ror registry data set cache

- update bi-weekly
- api lookup feature
- migration / install
- translations
  - english is ready
  - other languages have the keys with empty message

kmccurley commented on schemas/affiliation.json:
The schema lists ror as a required field. What happens when an author is affiliated with an institution that does not have an assigned ROR ID? While it's probably safe to assume that universities have these, it's certainly not true for industrial affiliations that are common in computer science. It took me only a few minutes looking in arxiv until I found a few such as https://arxiv.org/pdf/2410.08087 https://arxiv.org/pdf/2410.08081 https://arxiv.org/pdf/2410.08067 https://arxiv.org/pdf/2410.02445 https://arxiv.org/pdf/2410.07629

asmecher replied:
@kmccurley, the ROR will not be a required field; it will always be possible to add plain-text affiliations in order to accommodate users working with non-ROR research institutions.

GaziYucel replied:
Thanks @kmccurley, @asmecher for your feedback. As Alec said, ror is not a required field. I pushed a corrected version.

GaziYucel commented:
Hi,

Some notes. 

1. 
I have made some assumptions regarding business logic in several places. The ones I am uncertain about are the import/export classes: 
* plugins/importexport/native/filter/NativeXmlPKPAuthorFilter.php
* plugins/importexport/native/filter/PKPAuthorNativeXmlFilter.php

2. 
During an upgrade from an earlier version, the ROR cache is filled during the upgrade. After this a migration is done for all affiliations currently saved for the authors. First a match is searched for all ROR on an exact match for the names. After this all remaining is migrated as a non-ROR affiliation. 

3. 
There is an issue regarding the ROR data. Most affiliations have a name in one language, sometimes 2 -3 languages. Not all have a name in English. This is an issue regarding searching for a name in the backend. 
For now I look for an alternate name and show this if an English name is not there, with at the end the language. For example: "Technische Universiteit Delft [Dutch]". 

4.
The class in "classes/task/UpdateRorRegistryDataset.php" is responsible for updating the ROR cache. There are a lot of records without a language code. These are marked with "no_lang_code". I replace these with the locale code 'en'. 

5.
There is a question regarding the class PKP\API\v1\submissions\PKPSubmissionController.

If you look at the method getContributor, you see this at the end of this method. 

```
return response()->json(
    Repo::author()->getSchemaMap()->map($author),
    Response::HTTP_OK
);
```

I expected that the method "getSchemaMap()->map($author), also transforms "affiliations" to it's schema. But this is not happening, only author is mapped. Affiliations stays as in PHP, with "_data" and other keys still in it. 

I defined in the schema of author (schema/author.json) the following: 

```
"affiliations": {
	"type": "array",
	"description": "The scholarly institution this contributor is employed by or affiliated with.",
	"apiSummary": true,
	"readOnly": true,
	"items": {
		"type": "object",
		"$ref": "#/definitions/Affiliation"
	}
},
```
This is what I came up with, but this feels as a hack. 

```
$mappedAffiliations = [];
foreach ($author->getAffiliations() as $key => $affiliation) {
    $mappedAffiliations[] = Repo::affiliations()->getSchemaMap()->map($affiliation);
}

$author->setAffiliations(new LazyCollection($mappedAffiliations));
```

How should I do this to make affiliations also transform/map to it's schema?

GaziYucel commented:
Hi @bozana. This PR is ready for review. Thanks.

bozana commented:
Hi @GaziYucel, I am still reviewing the code and would also like to test it locally (also to understand some things better) but I wanted to already suggest something regarding the native import/export:
I think we could introduce the additional element `rorAffiliation` for the author, that could occur several times and that would contain the element `ror` and localized element `name`. 
The existing localized element `affiliation` could be used for the authors affiliations without ROR.
The examples how multiple affiliations and an affiliated institution with its identifier are represented in JATS can be found here: https://jats.nlm.nih.gov/archiving/tag-library/1.3/element/aff.html

bozana commented:
Hi @GaziYucel, 
and maybe regarding your last question above -- "And a question regarding the class PKP\API\v1\submissions\PKPSubmissionController" -- regarding the mapping of affiliations to the author: I think you could take a look how this is solved for issues and articles, e.g. here https://github.com/pkp/ojs/blob/main/classes/issue/maps/Schema.php#L111. I think something like this could work for this case as well.

bozana commented:
Hi @GaziYucel, 
And regarding the replacing `no_lang_code` with `en`: I think we cannot do it -- this way the original `en` entry can be overridden. We just cannot map `no_lang_code` to anything correctly. We would need to leave it as is.

bozana commented on classes/ror/Ror.php:
```diff
@@ -0,0 +1,26 @@
+class Ror extends DataObject
+{
+    public const STATUS_ACTIVE = 1;
+    public const STATUS_INACTIVE = 0;
```
I believe we could add the getters and setters -- I think we decided to use them so that we better support type hinting...

bozana reviewed (commented):
Hi @GaziYucel, I left some comments. 
Because we decided not to copy ROR names to author_affiliation_settings table some things will change here, so I can take a look once that is done...
About a few things I would maybe also need to think a little bit more, but I left a comment -- maybe you would immediately have an idea...
And because I have not taken a look at the OJS PR yet i.e. I would need some time to see how such affiliations need to be exported into different formats, I will probably first then know what Affiliation::getLocalized... function are needed... 
So I am already submitting this comments so that you can eventually already integrate some of them when working on the new version without copying...
I am wondering if we should keep this code version in case Alec has good arguments that we should proceed with copying -- because he has not said anything yet... although I do not think it will be the case...
Thanks a lot for this great work!

bozana commented on classes/author/Author.php:
```diff
@@ -256,4 +257,117 @@ public function getLocalizedCompetingInterests(): ?string
+     *
+     * @return string
+     */
+    public function getLocalizedAffiliationNamesAsString(?string $preferredLocale = null, ?string $separator = ', '): string
```
Lets use ';' as a default separator

GaziYucel commented on classes/author/Author.php:
```diff
@@ -256,4 +257,117 @@ public function getLocalizedCompetingInterests(): ?string
+     *
+     * @return LazyCollection
+     */
+    public function getAffiliations(): LazyCollection
```
I use methods which are in LazyCollection such as "all()".

bozana replied:
I am still not sure if this is the best way to go, but lets see it at the end...

GaziYucel replied:
I changed them to array.

GaziYucel commented on classes/ror/DAO.php:
```diff
@@ -0,0 +1,206 @@
+    /** @copydoc EntityDAO::update() */
+    public function update(Ror $ror): void
+    {
+        if (empty($ror->getId())) {
```
ror->getId is never empty/null. You are correct.

GaziYucel commented on classes/migration/install/AffiliationsMigration.php:
```diff
@@ -0,0 +1,60 @@
+            $table->bigInteger('author_affiliation_id');
+            $table->string('locale', 28)->default('');
+            $table->string('setting_name', 255);
+            $table->mediumText('setting_value')->nullable();
```
Yes, the value for name can be null. I add all current locales to the table, even if they are null.

bozana replied:
OK.
Note that you might need to get the possible submission locales differently, i.e. `$submission->getPublicationLanguages($context->getSupportedSubmissionMetadataLocales());`

GaziYucel replied:
Hi @bozana,
I can't figure out how to get $submission in the classes PKP\core\PKPApplication\FieldAffiliations. I do see this in PKP\components\forms\publication\ContributorForm, but it's always null. 

I do see different locales in PKPApplication::get()->getRequest()->getContext(): 
- supportedAddedSubmissionLocales
- supportedDefaultSubmissionLocale
- supportedFormLocales
- supportedLocales
- supportedSubmissionLocales
- supportedSubmissionMetadataLocales

But you pointed out that that is not correct. Thanks.

bozana replied:
Hi @GaziYucel, you need submission only because of allowed/supported locales, right?
If so, maybe: the ContributorForm actually contains all needed information about locales, so maybe to provide them (locales that you need for the affiliation field) here https://github.com/pkp/ui-library/blob/main/src/components/ListPanel/contributors/ContributorsListPanel.vue#L393, similar as you provide the authorId?

GaziYucel replied:
I already saw it there, but the variables are called different there, not sure if these have the values you wrote above. These are the locales in ContributorsListPanel:
- primaryLocale, e.g. 'en'
- supportedFormLocales, e.g. [ { key: 'en', label: 'English' }, { key: 'fr_CA', label: 'French (Canada)' } ]
- visibleLocales, e.g. [ 'en' ]
​
And the one I use now in PHP is $submissionContext->getSupportedSubmissionLocales()

Which one is the correct one than?

bozana replied:
The locales for the ContributorForm are provided here: https://github.com/pkp/pkp-lib/blob/main/classes/components/listPanels/ContributorsListPanel.php#L120. Those are the locales of the ContributorsListPanel.
And locales provided to the ContributorsListPanel (s. different places/handler in the code) are:
```
$locales = collect($submissionContext->getSupportedSubmissionMetadataLocaleNames() + $submission->getPublicationLanguageNames())
            ->map(fn (string $name, string $locale) => ['key' => $locale, 'label' => $name])
            ->sortBy('key')
            ->values()
            ->toArray();
```
Those are the supportedFormLocales.
The primary and visibleLocales are the same for each Form.
So I think you should only work with these locales for affiliations as well.
 
`$submissionContext->getSupportedSubmissionLocales() ` is not enough because they contain only currently enabled submission locales (and not metadata locales). The submission locales are those a submission can be published in, i.e. the article full text; and metadata locales can contain/allow additional locales for metadata -- e.g. an article in German, but the metadata also in English. 
Additionally, if already some metadata exists for a submission in a locale that is in the meantime disabled in journal settings -- lets say the metadata were entered in Finish, but then the Finish is disabled -- the users/editor still needs to be able to edit those existing metadata (in Finish). That is why we consider also $submission->getPublicationLanguageNames() (that returns all existing publication languages).

I have not taken a look into the ui-library, but will take a look where you need to locales and how...
But I believe those 3 kinds of form locales should be enough and correct for affiliations...

GaziYucel commented on classes/migration/install/RorsMigration.php:
```diff
@@ -0,0 +1,68 @@
+            $table->bigInteger('ror_id');
+            $table->string('locale', 28)->default('');
+            $table->string('setting_name', 255);
+            $table->mediumText('setting_value')->nullable();
```
Yes, the value for name can be null. I add all current locales to the table, even if they are null.

GaziYucel commented on classes/ror/maps/Schema.php:
```diff
@@ -0,0 +1,91 @@
+                    break;
+            }
+        }
+        $output = $this->schemaService->addMissingMultilingualValues($this->schema, $output, $this->context->getSupportedFormLocales());
```
You are right, we need to add the locales as they are provided by ROR. 
Line removed.

GaziYucel commented on controllers/grid/users/author/form/PKPAuthorForm.php:
```diff
@@ -225,7 +227,7 @@ public function execute(...$functionParams)
         $author->setFamilyName($this->getData('familyName'), null);
         $author->setPreferredPublicName($this->getData('preferredPublicName'), null);
-        $author->setAffiliation($this->getData('affiliation'), null); // localized
+        $author->setAffiliations($this->getData('affiliations'));
```
Yes, affiliations is always LazyCollection.

GaziYucel commented on classes/ror/Repository.php:
```diff
@@ -0,0 +1,170 @@
+     * @param Ror|null $object Ror being edited. Pass `null` if creating a new submission
+     * @param array $props A key/value array with the new data to validate
+     * @param array $allowedLocales The context's supported locales
+     * @param string $primaryLocale The context's primary locale
```
I removed the "validate" method. As you pointed out, it's not needed.

GaziYucel commented on classes/affiliation/maps/Schema.php:
```diff
@@ -0,0 +1,91 @@
+                    break;
+            }
+        }
+        $output = $this->schemaService->addMissingMultilingualValues($this->schema, $output, $this->context->getSupportedFormLocales());
```
Author has no submissionId in it's schema, only publicationId. I added get/setPublicationId. 
This is what I implemented: 
```
$author = Repo::author()->get($item->getAuthorId());
$submissionId = Repo::publication()->get($author->getPublicationId())->getData('submissionId');
$locales = Repo::submission()->get($submissionId)->getPublicationLanguages($this->context->getSupportedSubmissionMetadataLocales());
$output = $this->schemaService->addMissingMultilingualValues($this->schema, $output, $locales);
```

GaziYucel commented:
Hi @bozana
I implemented validation for affiliations in this commit: https://github.com/pkp/pkp-lib/pull/10460/commits/f92025f34b10251d3494a3f57179c0f72837d343

I had to remove Repo::affiliation()->saveAffiliations(..) from Repo::author()->dao->insert(..) and Repo::author()->dao->update(...) for this. I added Repo::affiliation()->saveAffiliations(..) to classes where the insert and update are called. Otherwise I can't check if an author exists. 

I put all in one commit so that you can review it more easily.

bozana commented on classes/affiliation/Affiliation.php:
```diff
@@ -0,0 +1,80 @@
+    /**
+     * Set name
+     */
+    public function setName(?array $name): void
```
I think we should stay consistent with other 'settings' method that could have locale and have something like:
```
    public function setName(string|array $name, ?string $locale = null)
    {
        $this->setData('name', $name, $locale);
    }

    public function getName(?string $locale = null)
    {
        return $this->getData('name', $locale);
    }
```
When you then provide an array as $name and no locale to setName() it will/should be set correctly.

bozana commented on classes/affiliation/DAO.php:
```diff
@@ -0,0 +1,175 @@
+                }
+                yield $row->author_affiliation_id => $fromRow;
+            }
+        });
```
This is not easy to read and understand, so maybe to do it similarly to author DAO and set affiliations?
To get the affiliation fromRow() and then get the Ror by ID and set the affiliations names.

bozana replied:
Then we would also not need that function getManyRorAsCollectionId() in Ror Collector

bozana commented on classes/affiliation/DAO.php:
```diff
@@ -0,0 +1,175 @@
+    public function fromRow(object $row): Affiliation
+    {
+        return parent::fromRow($row);
+    }
```
No need for this function then

bozana commented on classes/ror/Ror.php:
```diff
@@ -0,0 +1,53 @@
+
+class Ror extends DataObject
+{
+    public const int STATUS_ACTIVE = 1;
```
it should be `public const STATUS_ACTIVE = 1;` i.e. without the type `int`
here and below

bozana commented on classes/affiliation/Collector.php:
```diff
@@ -0,0 +1,161 @@
+    {
+        $this->authorIds = $authorIds;
+        return $this;
+    }
```
It seems this function is never used? 
Then maybe to leave the filtering only by one author ID?

bozana commented on classes/affiliation/Collector.php:
```diff
@@ -0,0 +1,161 @@
+    public ?string $name = null;
+
+    /** Get affiliations with a searchPhrase */
+    public ?string $searchPhrase = null;
```
It seems we do not consider this anywhere

bozana commented on classes/migration/upgrade/v3_5_0/I7135_CreateNewRorRegistryCacheTables.php:
```diff
@@ -0,0 +1,171 @@
+                function (JoinClause $join) {
+                    $join
+                        ->on('r_s.setting_value', '=', 'a_s.setting_value')
+                        ->on('r_s.locale', '=', 'a_s.locale')
```
does locale needs to be the same?

bozana commented on classes/migration/upgrade/v3_5_0/I7135_CreateNewRorRegistryCacheTables.php:
```diff
@@ -0,0 +1,171 @@
+                "id" => null,
+                "authorId" => $row->author_id,
+                "ror" => $ror->_data['ror'],
+                "name" => $ror->_data['name']
```
Maybe to use $ror->getData('ror') and $ror->getData('name')

bozana commented on classes/migration/upgrade/v3_5_0/I7135_CreateNewRorRegistryCacheTables.php:
```diff
@@ -0,0 +1,171 @@
+                "ror" => $ror->_data['ror'],
+                "name" => $ror->_data['name']
+            ];
+            $affiliation->setAllData($params);
```
Maybe to use $affiliation->setAuthorId(),  $affiliation->setRor() and $affiliation->setName()

bozana commented on classes/migration/upgrade/v3_5_0/I7135_CreateNewRorRegistryCacheTables.php:
```diff
@@ -0,0 +1,171 @@
+            ];
+            $affiliation->setAllData($params);
+
+            Repo::affiliation()->dao->updateOrInsert($affiliation);
```
In upgrade scripts we are using only SQLs and no objects and functions -- so that the upgrade can be correctly done for the given version even if we change the objects and code.
Thus, we should try to use the query builder to insert or update the affiliations.

bozana commented on classes/migration/upgrade/v3_5_0/I7135_CreateNewRorRegistryCacheTables.php:
```diff
@@ -0,0 +1,171 @@
+            ];
+            $affiliation->setAllData($params);
+
+            Repo::affiliation()->dao->updateOrInsert($affiliation);
```
Here as well, we should use the query builder to insert or update the affiliation

bozana commented on classes/affiliation/Repository.php:
```diff
@@ -0,0 +1,302 @@
+            }
+        }
+
+        $affiliation->setAllData($params);
```
I think it is better to use setName, setRor and get rid of $params array.

bozana commented on classes/affiliation/Repository.php:
```diff
@@ -0,0 +1,302 @@
+     */
+    public function getByAuthorId(int $authorId, ?string $submissionLocale = null): array
+    {
+        return iterator_to_array(
```
instead of using the php function iterator_to_array we should use Laravel all() function:
```
$this->getCollector()
  ->filterByAuthorId($authorId)
  ->getMany($submissionLocale)
  ->all()
```

Here and elswhere...

bozana commented on api/v1/submissions/PKPSubmissionController.php:
```diff
@@ -1558,6 +1559,15 @@ public function addContributor(Request $illuminateRequest): JsonResponse
+
+        $params['id'] = $newId;
+        $author->setId($newId);
+        $author->setAffiliations($params['affiliations']);
```
is $params['affiliations'] array of Affiliation objects?

bozana replied:
I think you should here do:
1. Repo::affiliation()->validate($author, $params, $submission, $submissionContext);
2. if error -> return
3. Repo::affiliation()->newDataObject($params);

Then no need to have author->setAffiliations, because you get the author at the end `$author = Repo::author()->get($newId);` which will get the affiliations.

bozana replied:
Maybe also to do here:
foreach ($params['affiliations'] as $affiliationParam) {
then the steps above
and having `Repo::affiliation()->validate(null, $params, $author, $submission, $submissionContext);` i.e. validating only one affiliation i.e. having the first parameter Affiliation object -- see below where the validate is implemented.

bozana replied:
Maybe to create Affiliation objects and use author->setAffiliations before adding the author.
Then in the author DAO the author is first added and then the affiliations saved, containing the author ID.

bozana commented on api/v1/submissions/PKPSubmissionController.php:
```diff
@@ -1558,6 +1559,15 @@ public function addContributor(Request $illuminateRequest): JsonResponse
         $author = Repo::author()->newDataObject($params);
         $newId = Repo::author()->add($author);
+
+        $params['id'] = $newId;
```
should this be 'authorId' instead of 'id'?

GaziYucel replied:
Here it is called 'id', it is however the author id.

bozana commented on classes/affiliation/Repository.php:
```diff
@@ -0,0 +1,302 @@
+            $validator = ValidatorFactory::make(
+                $props,
+                $schemaService->getValidationRules(PKPSchemaService::SCHEMA_AUTHOR, $allowedLocales)
+            );
```
I do not think you need this

bozana commented on classes/affiliation/Repository.php:
```diff
@@ -0,0 +1,302 @@
+        $allowedLocales = $submission->getPublicationLanguages($context->getSupportedSubmissionMetadataLocales());
+
+        // Check if author exists
+        if(!empty($props['id'])){
```
can we call it `authorId` instead of `id`?

bozana commented on classes/affiliation/Repository.php:
```diff
@@ -0,0 +1,302 @@
+                $schemaService->getValidationRules(PKPSchemaService::SCHEMA_AUTHOR, $allowedLocales)
+            );
+            $validator->after(function ($validator) use ($props) {
+                if (isset($props['id']) && !$validator->errors()->get('id')) {
```
If you check above for `isset($props['id'])` then no need here

bozana commented on classes/affiliation/Repository.php:
```diff
@@ -0,0 +1,302 @@
+     *
+     * @hook Affiliation::validate [[&$errors, $object, $props, $submission, $context]]
+     */
+    public function validate(?Author $author, array $props, Submission $submission, Context $context): array
```
Maybe to have the syntax as in all validate functions: to validate only one affiliation and to have the first parameter Affiliation object that could be null ?

bozana commented on classes/affiliation/Repository.php:
```diff
@@ -0,0 +1,302 @@
+     */
+    public function saveAffiliations(Author $author): void
+    {
+        $affiliations = $author->getAffiliations();
```
Here we can expect the affiliations to always be an array of Affiliation objects. (I commented above in PKPSubmissionController in add/editContributor)

GaziYucel replied:
There are many places where Repo::author()->add() and Repo::author()->edit() are called. If I don't do the check here, I expect that there will be unforeseen errors. That's why I added the extra check here.

bozana commented on classes/author/Author.php:
```diff
@@ -256,4 +272,72 @@ public function getLocalizedCompetingInterests(): ?string
+    }
+
+    /**
+     * Get the localized affiliations as an array.
```
localized affiliations names

bozana commented on plugins/importexport/native/pkp-native.xsd:
```diff
@@ -187,6 +187,14 @@
+                <complexType>
+                    <sequence>
+                        <element name="ror" type="string" minOccurs="0" maxOccurs="unbounded" />
+                        <element name="name" type="pkp:localizedNode" minOccurs="0" maxOccurs="unbounded" />
```
Should the minOccurs for ror and name be 1 -- if there is a rorAffiliation, then both elements should exist, no?

bozana commented on plugins/importexport/native/filter/NativeXmlPKPAuthorFilter.php:
```diff
@@ -106,13 +106,6 @@ public function handleElement($node)
-                            $locale = $publication->getData('locale');
-                        }
-                        $author->setAffiliation($n->textContent, $locale);
-                        break;
```
I believe this should stay, and the new element rorAffiliation needs to be considered as well, no?
Maybe the only thing is: if we have the affiliation element, if we would like to see if the ROR exists for it, like in the upgrade script :thinking:

bozana reviewed (commented):
Hi @GaziYucel, thanks a lot for the good work and improvements!
I left a few comments.
I believe, that having the ROR object in an affiliation would make it easier to deal with names in different locales, e.g. we will not need to worry about setting ROR default name to the affiliation name with the submission locale, and also getting getLocalizedAffiliations... should be easier. I am only not sure how that would work with the ui-library... Let me know what do you think...

bozana commented on classes/ror/Repository.php:
```diff
@@ -0,0 +1,120 @@
+    public function add(Ror $ror): int
+    {
+        $id = $this->dao->insert($ror);
+        Hook::call('Ror::add', [$ror]);
```
Before the hook I think we need to get the ror (by ID) again, because else the code using the hook will not bet the ror ID

bozana commented on controllers/grid/settings/user/form/UserDetailsForm.php:
```diff
@@ -182,7 +182,7 @@ public function initData()
             $data = [
                 'givenName' => $author->getGivenName(null), // Localized
                 'familyName' => $author->getFamilyName(null), // Localized
-                'affiliation' => $author->getAffiliation(null), // Localized
```
Hmmm... This file is used in the quickSubmit plugin. I do not think we can use the new affiliationField UI component there :-( I/we need to think how to solve it best...

bozana commented on controllers/grid/users/author/form/PKPAuthorForm.php:
```diff
@@ -225,7 +225,7 @@ public function execute(...$functionParams)
         $author->setGivenName(array_map('trim', $this->getData('givenName')), null);
         $author->setFamilyName($this->getData('familyName'), null);
         $author->setPreferredPublicName($this->getData('preferredPublicName'), null);
-        $author->setAffiliation($this->getData('affiliation'), null); // localized
```
Note for myself: For this file as well. It is only used in quickSubmit plugin and we need to see how to solve the new affiliations there for 3.5

bozana commented on controllers/grid/users/author/form/PKPAuthorForm.php:
```diff
@@ -225,7 +225,7 @@ public function execute(...$functionParams)
         $author->setFamilyName($this->getData('familyName'), null);
         $author->setPreferredPublicName($this->getData('preferredPublicName'), null);
-        $author->setAffiliation($this->getData('affiliation'), null); // localized
+        $author->setAffiliations($this->getData('affiliations'));
```
No need to change anything in this file -- it is deprecated.

blesildaramirez commented on classes/migration/upgrade/v3_5_0/I7135_CreateNewRorRegistryCacheTables.php:
```diff
@@ -0,0 +1,166 @@
+            $table->comment('Ror registry dataset cache');
+            $table->bigInteger('ror_id')->autoIncrement();
+            $table->string('ror')->nullable(false);
+            $table->string('display_locale', 28)->default('');
```
Hi @GaziYucel , I encountered an SQL error regarding this column having only a length of 28. In my case, the value being set is "no_lang_code: HUCH, HUCS, HUS, HYKS" which contains 35 characters, so my migration was unsuccessful for a specific record included in the records. Can you please check? Thanks!

GaziYucel replied:
Hi @blesildaramirez. Thank you for pointing this out. 
The schema here is actually the same as other objects in OJS. That column should be like that. As mentioned in the other issue FieldAutosuggest, there was some error in the import.

GaziYucel commented:
Hi @bozana,

I made a change in the database for rors/ror_settings. 

I have made more performance tests on my shared hosting. Not sure if my hoster did some downgrades in my hosting, but I found it less performant as my tests before for lookups of rors through the UI. This is due to several sql "like" lookups in rors.ror / ror_settings.setting_value. 

I added an extra column to rors, namely "search_phrase", where I concatenate the ROR iD (without https://ror.org/) and names during the import. The lookup is now done through a single like only on this column. 
With this change, I see a performance bump of 3 - 4 times (before 1,5s > after 0.2 - 0.5ms). The other side of the coin is that the table rors have grown from 20 to 30 MB.

GaziYucel commented on classes/ror/Ror.php:
```diff
@@ -0,0 +1,101 @@
+    /**
+     * Get cleaned name, e.g. remove NO_LANG_CODE, set required locale
+     */
+    public function getNameCleaned(?string $requiredLocale = null): mixed
```
I added this method to return the names with 'no_lang_code' cleaned. This is used in Affiliation::fromRow. I find this cleaner than mixing this in Affiliation Object.

bozana commented:
@asmecher, if I make ask you about that performance issue @GaziYucel mentions above -- what would be a good way to proceed/test/solve... ?
Thanks a lot!

asmecher commented:
@bozana and @GaziYucel, would a partial index on the `setting_value` solve the problem? I'd prefer this if it's feasible, as it doesn't require us to duplicate the data. We use this approach with `publication_settings` e.g. so that titles can be searchable.

https://github.com/pkp/pkp-lib/blob/c5ebc057b122ca01fa174e156734fdc7ae1dcb91/classes/migration/install/SubmissionsMigration.php#L90..L96

bozana commented:
Merged from here https://github.com/pkp/pkp-lib/pull/10880.
Thus closing
